### DISCUSSION TITLE
Solution build configurations ignored when using "dotnet build" and "dotnet run"

### DISCUSSION THREAD

**Discussion by PJB3005** [dotnet/msbuild#10336] [Q&A] [answered]
Hi, sorry if this is the wrong place to ask.

In our project's solution file we have more than the usual Debug & Release configurations, we also have "Tools" and "DebugOpt". These configurations are however exclusive to the main code of our project (so not used for any analyzers we also have in-tree). This means the solution config looks like this. Most things use "Tools" but some configurations are set to "Release":

![image](https://github.com/dotnet/msbuild/assets/8107459/486eecb6-a3a6-4282-9f9a-c2c145971675)

This works fine, however I just discovered that `dotnet run` and `dotnet build` completely ignore this when compiling a specific project. If you target a specific project, it will use the configuration of the **targeted** project for all dependencies, rather than the one set in the solution.

This causes loud compilation failures in our project on default config, but I am more concerned in general about the implications of this. It means that every time somebody uses `dotnet build` in our project, it's effectively using invalid configuration on many projects and just happening to not explode. This seems absurd to me.

is there anything I can specify here so that `dotnet run` is more in line with what my IDE does?

**Comment by KalleOlaviNiemitalo** [comment, marked as answer]
I can think of the following approaches, ordered by descending preference:

1. Add `SetConfiguration` metadata to the  `ProjectReference` item, to choose a configuration that is supported by the referenced project.  [Common MSBuild project items](https://learn.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2022#projectreference)
2. Run MSBuild on the solution rather than on an individual project, but specify the project as part of the target.  [Build specific targets in solutions by using MSBuild.exe](https://learn.microsoft.com/en-us/visualstudio/msbuild/how-to-build-specific-targets-in-solutions-by-using-msbuild-exe?view=vs-2022)
3. Add a custom target that runs after AssignProjectConfiguration and replaces unsupported configurations with supported ones.  This might be incompatible with MSBuild static graph.
4. Set properties to fake a solution build and cause AssignProjectConfiguration itself to choose the same configurations as in the solution.  I'm not sure this would work, because a real solution build also involves metaprojects that you wouldn't get this way.

**Reply by PJB3005** [reply]
I won't get around to trying it, but `SetConfiguration` definitely feels like the best solution there with our project configuration. Thanks.

**Comment by rainersigwald** [comment]
The core problem here is that solution configurations are used only when building in the context of a solution, while `dotnet run` operates on a _project_. I don't think there's a way to change that currently. ~I'm going to move this to SDK.~ I would move this to dotnet/sdk except discussions aren't enabled there. cc @baronfel

For `dotnet build` you should be able to use @KalleOlaviNiemitalo's suggestion 2 from https://github.com/dotnet/msbuild/discussions/10336#discussioncomment-9978103 to build the project in the solution context.

**Comment by matherm-aboehm** [comment]
I've written a workaround based on @KalleOlaviNiemitalo's suggestion 4. It also solves issues with restore operations, which is ignoring solution configuration even when building in the context of a solution.
See my [gist](https://gist.github.com/matherm-aboehm/96eaa496bc67c1ab1a1d022052f51ad0) for the code.
The files need to be placed in the same folder as the solution file.
